Zmon client

mohabusama opened the pull request:
- Add Zmon client
- Restructure commands
- Add tests

Note: some of the Zmon client method names are not consistent (update vs create vs add). It is somehow similar to the commands. I was not sure whether we should unify this in this iteration or wait for next upgrade, in order to avoid confusion for users (`zmon e push` vs `zmon e create`)

mohabusama committed:
Initial zmon client lib

mohabusama committed:
Add group & member methods

mohabusama committed:
Fix travis.yaml

mohabusama committed:
Fix flake8

mohabusama committed:
Restructure zmon-cli commands

Note: Not working yet. WIP

mohabusama committed:
Restructure tested client

+ Passing tests
+ Passing flake8

TODO: ZMON client tests

mohabusama committed:
Add Zmon client tests

WIP

mohabusama committed:
Add more tests

mohabusama committed:
Add helper functions

+ compare entities
+ get valid entity id

mohabusama committed:
Refactor to lazy load Zmon client

+ Pass CLI tests

mohabusama committed:
Fix alert def test

hjacobs commented on zmon_cli/__main__.py:
```diff
@@ -1,7 +0,0 @@
-from zmon_cli.main import main
```
why do you remove `__main__.py` ? this was convenience to run `python3 -m zmon_cli`

mohabusama replied:
Done.

hjacobs commented:
1. I would like to keep `__main__.py` to use `python3 -m zmon_cli` (that's how all STUPS tools work)
2. the output is mangled with "Action" output, i.e. I can't use standard UNIX tools anymore (like `awk`) --- please use the `--output` flag to suppress the "Action" output (this is automatically done with `OutputFormat`)

mohabusama committed:
Adjust commands output format and error handling

+ Add Output context manager, with optional printer (for custom table printing)
+ Accept text, yaml or json (with optional pretty print)
+ Add more deeplinks urls to Zmon client
+ Update tests

mohabusama commented:
Added custom `Output` to adjust unified output formatting and error/exception handling. This is done for most commands, and default format varies depending on the command.

hjacobs commented:
I get an KeyError now:
```
 $ python3 -m zmon_cli ent filter type local
 EXCEPTION OCCURRED: 'team'
Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/hjacobs/workspace/zmon-cli/zmon_cli/__main__.py", line 6, in <module>
    main()
  File "/home/hjacobs/workspace/zmon-cli/zmon_cli/main.py", line 10, in main
    cli()
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/click/decorators.py", line 27, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/hjacobs/workspace/zmon-cli/zmon_cli/cmds/entity.py", line 57, in filter_entities
    act.echo(entities)
  File "/home/hjacobs/workspace/zmon-cli/zmon_cli/output.py", line 95, in echo
    self.printer(out, self.output)
  File "/home/hjacobs/workspace/zmon-cli/zmon_cli/output.py", line 115, in render_entities
    rows.sort(key=lambda r: (r['id'], r['type'], r['team']))
  File "/home/hjacobs/workspace/zmon-cli/zmon_cli/output.py", line 115, in <lambda>
    rows.sort(key=lambda r: (r['id'], r['type'], r['team']))
KeyError: 'team'

mohabusama committed:
Remove team col from entities table output

mohabusama commented:
Removed `team` from col output. I assumed all entities has team assigned.

hjacobs commented:
@mohabusama you can actually introduce "last_modified" column instead (with nice "..ago" coloring by "print_table")

mohabusama committed:
Add Maodified col to entities print_table

hjacobs commented:
:+1:

mohabusama commented:
👍
